Mimetypedetector

rullzer opened the pull request:
As described in https://github.com/owncloud/core/issues/4774#issuecomment-88452605
Moving aways from the mimetype stuff in OC_Helper.

available unit tests are copied and modified.

It should not provide to much trouble as most functions were already wrappers around this.

rullzer committed:
Added mimetype detector

* Copied unit tests from old functions

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/80b3ca7f-b43d-4e58-906a-31f0743ff4b2) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/80b3ca7f-b43d-4e58-906a-31f0743ff4b2) completed: **25 new issues, 7 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/a3bde64c-6e9e-47fd-93b4-80b59215ed10) was created.

rullzer commented:
CC: @DeepDiver1975 @Xenopathic

Xenopathic commented on tests/lib/files/type/detection.php:
```diff
@@ -0,0 +1,91 @@
+
+	public function testGetSecureMimeType() {
+		$detection = new Detection();
+		$detection->registerTypeArray(include 'mimetypes.list.php');
```
`mimetypes.list.php` is used for custom mimetypes, so probably shouldn't be used here in the tests. Unless you want to try and tackle #15384 in this PR as well?

rullzer replied:
Ah you are right. I just made the old unit tests work again. I'll have a look later this week to make more proper unit tests.

MorrisJobke commented:
@rullzer Still valid? Then please rebase ;)

rullzer committed:
Move mimetypedetection to files/type/detection

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/079fc4a5-4b17-469a-bf67-3a36d9768f65) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/a416e1cf-a2c4-4879-9a30-b9dbf8fc0fb3) was created.

rullzer commented:
Rebased, moved all the mimetype related stuff to OC/Files/Type/Detection (OCP/Files/IMimetypeDetector).

Test still work. Altough I might extend them.
Old functions call the fancy new mimetypedetector. Just removing them is tricky since some (official) apps depend on them... even tough they are private.

Lets see what jenkins has to say.

rullzer commented:
@owncloud-bot retest this please

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/e4ba50ad-f410-4fbf-980d-23b427f23cc7) was created.

rullzer commented:
@owncloud-bot retest this please

Xenopathic commented on lib/private/helper.php:
```diff
@@ -496,9 +428,10 @@ static function getSecureMimeType($mimeType) {
 	 *
 	 * @param string $data
 	 * @return string
+	 * @deprecated Use \OC::$server->getMimeTypeDetector->detectString($data)
```
`getMimeTypeDetector()`

rullzer replied:
Thanks fixed

Xenopathic commented on lib/private/server.php:
```diff
@@ -444,6 +446,21 @@ public function __construct($webRoot) {
 			return new \OC\Files\Mount\Manager();
 		});
+		$this->registerService('MimeTypeDetector', function(Server $c) {
+			$mimeTypeDetector = new Detection();
```
Given the size of this file, I'd suggest using a fully qualified class name here and avoiding the `use` above.

rullzer replied:
Yes definitely.. was a remainder of the old code I wrote :) Thanks fixed.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/e572f10e-2a5d-447e-96b5-3dbf32ca7441) was created.

rullzer commented:
@DeepDiver1975 this should fix #17421

As you can see I deprecated a lot of functions in OC_Helper. Now I know that we can usually just remove private functions but this was used in a lot of places since there was no OCP alternative. What to do in that case?

rullzer commented:
Review time!

rullzer commented:
@owncloud-bot retest this please

Xenopathic commented on lib/private/files/type/detection.php:
```diff
@@ -170,4 +178,68 @@ public function getSecureMimeType($mimeType) {
+
+		// Is it a dir?
+		if ($mimetype === 'dir') {
+			$this->mimetypeIcons[$mimetype] = \OC::$server->getURLGenerator()->imagePath('core', 'filetypes/folder.png');
```
Can we use dependency injection for the URL generator?

rullzer replied:
Yes, much cleaner.
Would also allow unit tests next to the integration tests

Xenopathic commented on lib/private/files/type/detection.php:
```diff
@@ -170,4 +178,68 @@ public function getSecureMimeType($mimeType) {
+				$custom = get_object_vars(json_decode(file_get_contents(\OC::$configDir . '/mimetypealiases.json')));
+				$this->mimeTypeAlias = array_merge($this->mimeTypeAlias, $custom);
+			}
+		}
```
It's weird that the mappings are loaded in the server container, while these aliases are loaded here in the class. Can we make this more consistent? Either everything in the constructor, or everything in the server container, not this split.

rullzer replied:
You are right. Maybe the constructor of the Detection class makes the most sense?

rullzer replied:
Altough just in the function is maybe easier since that makes sure it is really only loaded when required...

Xenopathic replied:
I like the new `addMappings()` function (although it should be really named `loadMappings()`), but can you do the same for this? Split the loading of mime aliases into its own function, to make this function less complex.

rullzer committed:
Loading of mapping/aliases is done in class

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/2f37117e-6330-4c53-870c-e31bccbf9a6a) was created.

rullzer commented:
@owncloud-bot retest this please

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/f2caebd2-9ccb-40e7-8dbb-a71f1feb77d5) was created.

rullzer commented:
@Xenopathic you are right. It looks cleaner now.

rullzer commented:
Lets see if Jenkins is still happy

rullzer commented:
@owncloud-bot retest this please

Xenopathic commented:
Looking good. Could you squash some of those commits though? We have 8 commits here, most of them fixes or adjustments of earlier ones.

rullzer commented:
@Xenopathic yeah will do. Just waiting for Jenkins to be finished.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/49d305a8-aca0-4e42-beee-28b1055abbed) was created.

rullzer commented:
Ok rebased and squashed... let see

rullzer commented:
@owncloud-bot retest this please

rullzer commented:
rebased... lets see if docker does its thing now.

rullzer commented:
@owncloud-bot retest this please

MorrisJobke commented:
@owncloud-bot retest this please

MorrisJobke commented:
Jenkins PR #17718

rullzer commented:
Jenkins is happy... see #17718 
So review :)

Xenopathic commented on lib/private/files/type/detection.php:
```diff
@@ -166,8 +226,65 @@ public function detectString($data) {
+		}
+		if ($mimetype === 'dir-external') {
+			$this->mimetypeIcons[$mimetype] = $this->urlGenerator->imagePath('core', 'filetypes/folder-external.
+png');
```
Bad line wrapping here - turn off auto-wrap in your editor :laughing:

rullzer replied:
Ai! This shows that we actually need unit tests for this.

Xenopathic commented on lib/private/helper.php:
```diff
@@ -465,9 +394,10 @@ static public function getFileTemplateManager() {
 	 *
 	 * @param string $path
 	 * @return string
+	 * @deprecated Use \OC::$server->getMimeTypeDetector()->detectPath($path)
```
Mention the deprecation version for this and all other functions lower down? The ones above are correct.

Xenopathic commented:
:+1: for the code, just need the whitespace and phpdoc fixes

rullzer committed:
Proper deprecate methods

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/3992fd2d-912e-487f-a0b6-7acb02d41e38) was created.

rullzer commented:
It seems unit testing the functions is not so trivial in the current setup (you can't set the vars to something you want easily since they read from a hard coded path). Let me think about how to solve that in a clean way.

Xenopathic commented:
@rullzer How about making the `loadMappings()` and `loadAliases()` functions public, and take a path/two paths to the JSON files. These functions then get called during instantiation in the server container, with the correct paths. In unit testing, we can call the functions with whatever paths we like, or not at all (defaults to empty array if not loaded, right?). Unfortunately you lose the deferred loading this way...

Alternatively, pass the config directory path to the constructor? That's how `\OC\Config` does it.

rullzer commented:
@Xenopathic I was thinking about the latter one (so passing the config dir to the constructor). That way it is easily testable with a temp dir that is passed. And the full class can actually be properly unit tested.

I'll fix that hopefully this evening.

rullzer committed:
Pass config dir as parameter to detection class

In order to properly test the mimetype function:

* constructor takes path to configdir
* Added unit tests for mimetype (only if vfsStream is available)

rullzer commented:
Unit tests are added. But depend on https://github.com/owncloud/3rdparty/pull/185.

MorrisJobke commented:
@rullzer Rebase needed

rullzer commented:
@MorrisJobke will do once the 3rdparty is merged and bumbed on master.

MorrisJobke commented:
> @MorrisJobke will do once the 3rdparty is merged and bumbed on master.

Makes sense :smiley:

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/748a46a2-9800-451a-80f6-ce3990d723d9) was created.

rullzer commented on tests/lib/files/type/detection.php:
```diff
@@ -0,0 +1,283 @@
+	public function testMimeTypeIconStaticIcons() {
+		if (!class_exists('org\\bovigo\\vfs\\vfsStream')) {
+			$this->markTestSkipped('Pacakge vfsStream not installed');
+		}
```
Only run this test if the vfsStream pacakge is avaialble.. since the outcome of https://github.com/owncloud/3rdparty/pull/185 is unclear

Xenopathic commented:
@owncloud-bot retest this please

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ab210fda-53b2-4cbe-87f2-c5ec01b32ab7) was created.

rullzer commented:
Rebased and added conditional to only run the test requiring vfsStream if it is available.

Review time!

rullzer commented:
@owncloud-bot retest this please

rullzer commented:
Jenkins is happy! #17889 
Lets get this in!

Review time: @MorrisJobke @PVince81 @DeepDiver1975 

@Xenopathic your :+1: is still valid I assume?

Xenopathic commented on lib/private/helper.php:
```diff
@@ -496,9 +428,10 @@ static function getSecureMimeType($mimeType) {
 	 *
 	 * @param string $data
 	 * @return string
+	 * @deprecated Use \OC::$server->getMimeTypeDetector()->detectString($data)
```
Missing deprecation version

rullzer replied:
Will add (right before  merge) else jenkins migt die...

Xenopathic commented on lib/public/files/imimetypedetector.php:
```diff
@@ -0,0 +1,70 @@
+// This means that they should be used by apps instead of the internal ownCloud classes
+namespace OCP\Files;
+
+interface IMimeTypeDetector {
```
Put a `@since` annotation on the whole interface (a brief description of the interface would be nice too)

MorrisJobke replied:
Oh nooo .... my script that should check for this, doesn't triggered :( Let me fix this :)

rullzer replied:
Yes... I blame @MorrisJobke here ;)
Anyway fixed locally. Will push in the end not to confuse the bot anymore than it has to ;)

Xenopathic commented on tests/lib/files/type/detection.php:
```diff
@@ -0,0 +1,283 @@
+
+		$result = $detection->detect($dir."/testimage.png");
+		$expected = 'image/png';
+		$this->assertEquals($expected, $result);
```
Pro-tip: you can turn these sorts of repeated tests into a single test with a data provider. Makes it really easy to add test cases in the future.

rullzer replied:
True, that test was largly copied from the existing test. But you are right :)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/916b3bda-6117-4f0d-898a-22aed8824a7f) was created.

Xenopathic commented:
@owncloud-bot retest this please

Xenopathic commented on lib/private/files/type/detection.php:
```diff
@@ -32,9 +36,28 @@
+	private $configDir;
+
+	/**
+	 * @param \IURLGenerator $urlGenerator
```
Err, this most *definitely* isn't `\IUrlGenerator` :laughing:

rullzer replied:
fixed

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/7b71f450-052c-4f1f-b710-3da9b427c3b0) was created.

rullzer commented:
@owncloud-bot retest this please

Xenopathic commented:
:+1: Let's go! cc @MorrisJobke @PVince81

MorrisJobke commented:
Code looks good and it still works :+1:
